File: rock/sdk/model/server/traj.py
"""Trajectory record + replay for the chat/completions proxy.

Two halves around the same JSONL schema (one record per line):

- :class:`TrajectoryRecorder` — invoked by the forward path after each upstream
  call (success or failure). Appends a small dict with
  ``request`` / ``response`` / ``status`` / ``response_time`` / ``model`` /
  ``stream``, and reports OTLP RT/count metrics. Stores responses verbatim
  (provider-specific fields like ``reasoning_content`` survive); for streaming
  calls ``response`` is the aggregated final ChatCompletion produced by
  ``ChatCompletionStreamState.get_final_completion().model_dump()``.

- :class:`SequentialCursor` — loads a JSONL trajectory once at startup;
  ``await cursor.next(expected_model=...)`` hands out the next record (full
  payload dict) and advances. Going past the end raises
  :class:`TrajectoryExhausted` so the proxy can return a clean 404.
"""

from __future__ import annotations

import asyncio
import json
import os
from pathlib import Path
from typing import Any

from rock.logger import init_logger
from rock.sdk.model.server.utils import (
    MODEL_SERVICE_REQUEST_COUNT,
    MODEL_SERVICE_REQUEST_RT,
    _get_or_create_metrics_monitor,
)

logger = init_logger(__name__)


# ---------------------------------------------------------------------------
# Recorder
# ---------------------------------------------------------------------------


class TrajectoryRecorder:
    """Appends one JSONL line per chat/completions call and reports OTLP metrics."""

    def __init__(self, traj_file: str | os.PathLike) -> None:
        self.traj_file = Path(traj_file)
        self.traj_file.parent.mkdir(parents=True, exist_ok=True)
        self._lock = asyncio.Lock()
        self._monitor = _get_or_create_metrics_monitor()

    async def record(
        self,
        *,
        request: dict[str, Any],
        response: dict[str, Any] | None,
        status: str,
        start_time: float,
        end_time: float,
        error: str | None = None,
    ) -> None:
        rt_seconds = end_time - start_time
        payload = {
            "model": request.get("model"),
            "stream": bool(request.get("stream")),
            "status": status,
            "response_time": rt_seconds,
            "start_time": start_time,
            "end_time": end_time,
            "request": request,
            "response": response,
            "error": error,
        }

        line = json.dumps(payload, ensure_ascii=False, default=str) + "\n"
        async with self._lock:
            await asyncio.to_thread(self._write_line, line)

        attrs = {
            "type": "chat_completions",
            "status": status,
            "sandbox_id": os.getenv("ROCK_SANDBOX_ID", "unknown"),
        }
        self._monitor.record_gauge_by_name(MODEL_SERVICE_REQUEST_RT, rt_seconds * 1000.0, attributes=attrs)
        self._monitor.record_counter_by_name(MODEL_SERVICE_REQUEST_COUNT, 1, attributes=attrs)

    def _write_line(self, line: str) -> None:
        with self.traj_file.open("a", encoding="utf-8") as f:
            f.write(line)


# ---------------------------------------------------------------------------
# Replay cursor
# ---------------------------------------------------------------------------


class TrajectoryExhausted(Exception):
    """Raised by ``SequentialCursor.next`` when all recorded steps have been served."""

    def __init__(self, position: int, total: int) -> None:
        super().__init__(f"trajectory exhausted at step {position} (total recorded steps={total})")
        self.position = position
        self.total = total


class SequentialCursor:
    """Hands out trajectory records one at a time, in recorded order."""

    def __init__(self, records: list[dict]) -> None:
        self.records = records
        self._idx = 0
        self._lock = asyncio.Lock()

    @classmethod
    def load(cls, path: str | os.PathLike) -> SequentialCursor:
        path = Path(path)
        if not path.is_file():
            raise FileNotFoundError(f"traj file not found: {path}")

        records: list[dict] = []
        with path.open("r", encoding="utf-8") as fp:
            for line in fp:
                line = line.strip()
                if not line:
                    continue
                records.append(json.loads(line))

        logger.info(f"[traj-replay] loaded {len(records)} record(s) from {path}")
        return cls(records)

    async def next(self, expected_model: str | None = None) -> dict:
        async with self._lock:
            if self._idx >= len(self.records):
                raise TrajectoryExhausted(position=self._idx, total=len(self.records))
            record = self.records[self._idx]
            self._idx += 1
            current_idx = self._idx - 1

        if expected_model:
            recorded_model = record.get("model")
            if recorded_model and recorded_model != expected_model:
                logger.warning(
                    f"[traj-replay] step {current_idx} model mismatch: "
                    f"recorded={recorded_model!r} requested={expected_model!r}"
                )
        return record

    def reset(self) -> None:
        self._idx = 0

    @property
    def position(self) -> int:
        return self._idx

    @property
    def total(self) -> int:
        return len(self.records)

File: rock/ts-sdk/src/model/server/traj.ts
/**
 * Trajectory record + replay for the chat/completions proxy.
 *
 * Two halves around the same JSONL schema (one record per line):
 *
 * - TrajectoryRecorder — invoked by the forward path after each upstream
 *   call (success or failure). Appends a small dict with
 *   request / response / status / response_time / model / stream.
 * - SequentialCursor — loads a JSONL trajectory once at startup;
 *   cursor.next(expectedModel=...) hands out the next record (full
 *   payload dict) and advances.
 *
 * Mirrors rock/sdk/model/server/traj.py.
 */

import * as fs from 'fs';
import * as path from 'path';
import { initLogger } from '../../logger.js';

const logger = initLogger('rock.model.server.traj');

// ---------------------------------------------------------------------------
// TrajectoryExhausted
// ---------------------------------------------------------------------------

/**
 * Raised by SequentialCursor.next when all recorded steps have been served.
 */
export class TrajectoryExhausted extends Error {
  readonly position: number;
  readonly total: number;

  constructor(position: number, total: number) {
    super(
      `trajectory exhausted at step ${position} (total recorded steps=${total})`,
    );
    this.name = 'TrajectoryExhausted';
    this.position = position;
    this.total = total;
  }
}

// ---------------------------------------------------------------------------
// TrajectoryRecorder
// ---------------------------------------------------------------------------

/** Parameters for TrajectoryRecorder.record(). */
export interface TrajectoryRecordParams {
  request: Record<string, unknown>;
  response: Record<string, unknown> | null;
  status: string;
  startTime: number;
  endTime: number;
  error?: string | null | undefined;
}

/**
 * Appends one JSONL line per chat/completions call.
 */
export class TrajectoryRecorder {
  private trajFile: string;

  constructor(trajFile: string) {
    this.trajFile = trajFile;

    // Create parent directory if it doesn't exist
    const dir = path.dirname(trajFile);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  }

  /**
   * Record a trajectory entry.
   *
   * Appends a JSON line to the traj file containing request, response,
   * timing, and status information.
   */
  async record(params: TrajectoryRecordParams): Promise<void> {
    const { request, response, status, startTime, endTime, error } = params;
    const rtSeconds = endTime - startTime;

    const payload: Record<string, unknown> = {
      model: request.model,
      stream: Boolean(request.stream),
      status,
      response_time: rtSeconds,
      start_time: startTime,
      end_time: endTime,
      request,
      response,
      error: error ?? null,
    };

    const line = JSON.stringify(payload) + '\n';

    // Synchronous file write wrapped in async for the interface
    await new Promise<void>((resolve, reject) => {
      try {
        fs.appendFileSync(this.trajFile, line, 'utf-8');
        resolve();
      } catch (e) {
        reject(e);
      }
    });
  }
}

// ---------------------------------------------------------------------------
// SequentialCursor
// ---------------------------------------------------------------------------

/**
 * Hands out trajectory records one at a time, in recorded order.
 */
export class SequentialCursor {
  private records: Record<string, unknown>[];
  private _idx: number;

  constructor(records: Record<string, unknown>[]) {
    this.records = records;
    this._idx = 0;
  }

  /**
   * Load a SequentialCursor from a JSONL trajectory file.
   */
  static load(filePath: string): SequentialCursor {
    if (!fs.existsSync(filePath)) {
      throw new Error(`traj file not found: ${filePath}`);
    }

    const records: Record<string, unknown>[] = [];
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n');

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) {
        continue;
      }
      try {
        records.push(JSON.parse(trimmed));
      } catch {
        // Skip malformed lines silently
      }
    }

    logger.info(
      `[traj-replay] loaded ${records.length} record(s) from ${filePath}`,
    );
    return new SequentialCursor(records);
  }

  /**
   * Return the next trajectory record and advance the cursor.
   *
   * @param expectedModel - If provided, logs a warning when the recorded
   *   model differs from the expected model (but does not throw).
   * @returns The next record.
   * @throws TrajectoryExhausted when all records have been consumed.
   */
  async next(expectedModel?: string): Promise<Record<string, unknown>> {
    if (this._idx >= this.records.length) {
      throw new TrajectoryExhausted(this._idx, this.records.length);
    }

    const record = this.records[this._idx]!;
    this._idx += 1;
    const currentIdx = this._idx - 1;

    if (expectedModel) {
      const recordedModel = record.model;
      if (recordedModel && recordedModel !== expectedModel) {
        logger.warn(
          `[traj-replay] step ${currentIdx} model mismatch: ` +
            `recorded=${JSON.stringify(recordedModel)} requested=${JSON.stringify(expectedModel)}`,
        );
      }
    }

    return record;
  }

  /** Reset the cursor back to the beginning. */
  reset(): void {
    this._idx = 0;
  }

  /** Current position (number of records consumed). */
  get position(): number {
    return this._idx;
  }

  /** Total number of loaded records. */
  get total(): number {
    return this.records.length;
  }
}
